### ISSUE TITLE
Property initializer with dependent property does not warn of possible null reference.

### ISSUE THREAD

**Issue by JoasE** [dotnet/roslyn#65658]
**Version Used**: 11

**Steps to Reproduce**:
Create a class with 2 required init properties.
Define a backing filed for one of the properties and reference the other property in the init body.
Building the code below does not show a possible null reference warning.
Running the code throws a NullReferenceException in the `init => _weirdProperty = Test.Name + value;`

```csharp
var test = new TestRequiredInit
{
    WeirdProperty = "a",
    Test = new Test { Name = "b" }
};

public class TestRequiredInit
{
    private readonly string _weirdProperty;

    public required Test Test { get; init; }

    public required string WeirdProperty
    {
        get => _weirdProperty;
        init => _weirdProperty = Test.Name + value;
    }
}

public class Test
{
    public required string Name { get; init; }
}
```
https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAbgIYBO6ALgKYDOXdAF50AOx4B3dABV+XAEo8AjgFcAlux7AAkqLVdGTIgSRGz6AOo8NwAArsA9gAce7LgE9h6AEStvAGkNzIlkBL3EpUME8dAA5VgBbHi9vACNvdABfIKZMgG4g7KQi5GQ0TBgZOUVVDS1dfWQTJid2NTZedE1WYAdRABtPKDgABnQAfQlrdjtHFzd3AqQg8s1azWAqsKj8dABzHi489DU9I6zS02JV5XUNzFHLadnnVw8g5vMDwSEAPgmpjZ7K8Fktgqd9MJ/pNnsD5h4vFEAHTxJLoMDoNj9FQ8MHEIolFCkSpRJorDBrO5aB5jVHJGLfY4Q84EoA==

**Diagnostic Id**: **MISSING**: Warning CS8602 Dereference of a possibly null reference.

**Expected Behavior**: Compiler shows warning on `_weirdProperty = **Test**.Name + value;`.
As we can not guarantee the order of the property initializers, `Test` can be null at the time of the `WeirdProperty` initializer.

**Actual Behavior**: No warning is shown.

**Comment by DoctorKrolic** [comment]
1. The `Name` property of the `Test` class isn't marked as nullable string (`string?`)
2. It is absolutely legal to concatenate `null` with some string (even the null `string`). For instance, try to run this program:
```cs
Console.WriteLine(null + (string?)null);
```
You will get empty output, not the null ref exception crash

3. Having your code written the way, when the order of properties in initializers changes the state of the program is an antipattern on its own. Don't write code this way.

**Comment by DoctorKrolic** [comment]
The `Test` property isn't marked as nullable. When you add the `?`, the warning is there: https://sharplab.io/#v2:C4LgpgHgDgNghgSwHYBoAmIDUAfAAgJgEYBYAKFwGYACAqgYSoG8yrWbrcAWKgWQAoAlExZtRANzgAnKsDABnYFQC8VJGADuVACrzgAJTABHAK4JJYNAEkkCYCNGtmpBy6oB1MGbQAFSQHsoMElgAE9lKgAiOAiUe1dWHQVwtU1ExUYqADk4AFswcIiAIwiqAF840VKAbjjy0jqyMkoafG1dAxMzC2tbMidRKEkECVkqczg0PyQYMNxCAAYqAH11T0kff0DgkJrSOObzTvM0NoUAflP0qgBzMGAqqmRbB4bnNgOjU2OaBfc1jYCQVCcX6rluiiUAD5lqsvL5AdtdvEnhDoSt/vCtqFwmkAHTZPJUTBUCQwYxgJFsOqvZq0NJ9fbUQ5fCw/RYE/IZcEPFEvMilIA=

**Comment by JoasE** [comment]
@DoctorKrolic I understand that. The issue is that it can be null during the `init` phase of the class. For example, when doing the below, the compiler does show a warning about the reference being null:
```csharp
public class TestCtorWarning
{
    private readonly string _weirdProperty;
    private readonly string _name;
    
    public TestCtorWarning(string weirdProperty, string name)
    {
        _weirdProperty = _name.Length + weirdProperty;
        _name = name;
    }
}
```
https://sharplab.io/#v2:C4LgpgHgDgNghgSwHYBoAmIDUAfAAgJgEYBYAKFwGYACAqgYSoG8yrWbrcAWKgWQAoAlExZtRI1gF8yU0mUo18VACpgAzsDrAA9gCcA6nB1JkAczLNSoqDoQA3OMDBUdYOGi1IYATxqEADFQA+gDuYAg6aAAKOlpQYDrAXgDc4lTWdg5OLm4e3r4BgUhwALZgKZasqanyKuqaugZGpny4/lSh4VExcQleKPlURaUCqRaioiFhEdGx8YlUALxBQ2AAdAAyYEgmwAAWVJjtU12zveXjbIUlTksr52wyEkA

The same warning should be shown on `_weirdProperty = **Test**.Name + value;` as we can not guarantee `Test` already has a value in the `init` phase. `Test` does however always have a value after the `init` phase, so it shouldn't be marked with a `?`.

In short: `init` bodys that reference non-null instance members should give a warning, unless the member is set in the constructor.

**Comment by JoasE** [comment]
I understand that this might not be the best code. But it still could happen, and a warning could prevent users from writing even worse code (one where `NullReferenceException`'s could occur.) or help them understand why their code isn't very good.

**Comment by RikkiGibson** [comment]
> In short: `init` bodys that reference non-null instance members should give a warning, unless the member is set in the constructor.

What you're asking for here is an interprocedural flow analysis. Unfortunately, this is something we have ruled out in general because of the computational complexity of doing so.

We might be able to make small adjustments to heuristics which are self-contained to set/init accessors. But this will require some investigation of the impact (i.e. potential of adding warnings to existing code) and a more formal description of what the initial nullable state of fields is in a property setter and why.

**Comment by weltkante** [comment]
I don't think "interprocedural flow analysis" is the primary point here. The order in which `init` accessors are called upon is arbitrary (depends on the call site) so all init bodies should treat their non-nullable variables as potentially having null values - just like in a constructor.

Asking for members being set in the constructor to be excluded from this treatment would be a bonus and not the main point requested.

Really the language should err on the side of safety, like it does in every other place. Not emitting a warning here seems like a bug given you don't want to do interprocedural flow analysis.

**Comment by RikkiGibson** [comment]
The language makes many simplifying assumptions for usability. For example, the following produces no warnings, and gives an NRE at runtime.

```cs
_ = new C();

class C
{
    string _field;
    public C()
    {
        M();
        _field = "a";
    }

    void M()
    {
        _field.ToString(); // no warning
    }
}
```

This is because we think that usually you are going to call methods after construction, but we don't want to stop you from calling methods during construction.

> all init bodies should treat their non-nullable variables as potentially having null values - just like in a constructor.

Certainly, init/set accessors are more on the side of "you're going to call this when not everything is initialized yet". We just have to investigate our options and the impact thoroughly. We would then have to conclude that the utility of additional warnings in init bodies is worth the breaks to existing code.

In particular, I think we would be introducing pain when a property setter assigns to some nested member. 

```cs
class Inner
{
    public string? Prop { get; set; }
}

class C
{
    public Inner _inner;
    public C(Inner inner) => _inner = inner;
    public string? Prop
    {
        get => _inner.Prop; // ok
        set => _inner.Prop = value; // new warning: 'inner' could be null
    }
}
```

**Comment by jcouv** [comment]
If we wanted to warn here, I think we'd have to say that within an `init`, the values of all other init-only properties may be null.
@RikkiGibson What do you think?

**Comment by RikkiGibson** [comment]
> If we wanted to warn here, I think we'd have to say that within an `init`, the values of all other init-only properties may be null. @RikkiGibson What do you think?

It sounds like the idea is that since an `init` is only ever called "before construction is complete", unlike a setter which can be called any time, it would be helpful to use a more "pessimistic" flow state for it.

I feel a bit shaky about it. It feels like you could keep people out of trouble in some cases if they happened to be using init properties, but if they changed to use set, we'd start giving them a more "optimistic" flow state, even if none of their call sites actually changed. It's the same situation if someone declares an `Initialize()` method and calls it from their constructor. If we help people in this way in one kind of member but not another, it could create an expectation that the help will be present in general, and an unpleasant surprise when the user finds out that it's not.

I think my overall feeling is, once construction becomes more complex than "run the base constructor, then assign all the current instance fields", things just get dicey. It's a path where we cannot provide many guardrails. This is why, in particular, roslyn-analyzers has a rule which warns when calling a virtual method from a constructor--there's too much risk of the override assuming the instance is initialized.

I think it might even be reasonable for the platform to provide an analyzer which encourages people to stick to assigning fields and auto-properties directly in their constructors, with some pattern (`this!.Initialize()` or similar) to say, "yeah, I know this could cause me to read unexpected nulls, but I'm OK with that".
